Update algorithms

npm1 opened the pull request:
Lands https://github.com/fedidcg/FedCM/pull/214 including followup comments


<!--
    This comment and the below content is programmatically generated.
    You may add a comma-separated list of anchors you'd like a
    direct link to below (e.g. #idl-serializers, #idl-sequence):

    Don't remove this comment or modify anything below this line.
    If you don't want a preview generated for this pull request,
    just replace the whole of this comment's content by "no preview"
    and remove what's below.
-->
***
<a href="https://pr-preview.s3.amazonaws.com/npm1/FedCM/pull/256.html" title="Last updated on May 27, 2022, 9:57 PM UTC (5464ecc)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/fedidcg/FedCM/256/b439d86...npm1:5464ecc.html" title="Last updated on May 27, 2022, 9:57 PM UTC (5464ecc)">Diff</a>

npm1 commented:
@samuelgoto PTAL

npm1 commented:
Ping

samuelgoto commented on spec/index.bs:
```diff
@@ -953,24 +973,32 @@ an internal [[#browser-api-state-machine]].
-         [=logged in=] or (by default) [=logged out=].
+    ::  Keeps track of whether the user has an open or closed session.
+         Can be [=loggedin=] or [=loggedout=] (by default).
+    :   <dfn>Prompt State</dfn>
```
When is this used? Can we remove it until we actually implement the other prompting styles? 

I think it will make this PR easier to merge if we limit it to things that have already been implemented, because we don't have to think hard about "what's the right thing to do" but rather just codify "what has been done / implemented".

samuelgoto commented on spec/index.bs:
```diff
@@ -1237,21 +1193,115 @@ dictionary FederatedTokens {
+To <dfn>sign-up</dfn> the user:
+    1. Let |metadata| be the result of running the [=fetch the client metadata=]
+        algorithm.
+    1. Display the |metadata|["privacy_policy_url"] link
```
We recently made this optional.

If |metadata|["privacy_policy_url"] is defined, then display ...

npm1 replied:
Done

samuelgoto commented on spec/index.bs:
```diff
@@ -1237,21 +1193,115 @@ dictionary FederatedTokens {
+    1. Display the |metadata|["privacy_policy_url"] link
+    1. If |metadata|["terms_of_service_url"] is defined then display the
+        |metadata|["terms_of_service_url"] link
+    1. Prompt the user to gather explicit intent to create an account
```
Should we also update this algorithm to include the branding JSON options, like the button color and logos?

npm1 replied:
Done

samuelgoto commented on spec/index.bs:
```diff
@@ -767,10 +788,10 @@ Sec-FedCM-CSRF: ?1
 <dl dfn-type="argument" dfn-for="client_metadata_endpoint_response">
-    :   <dfn>privacy_policy_url</dfn> (required)
-    ::  A link to the [=RP=]'s privacy policy.
+    :   <dfn>privacy_policy_url</dfn> (optional)
```
thanks!

samuelgoto commented on spec/index.bs:
```diff
@@ -953,24 +973,29 @@ an internal [[#browser-api-state-machine]].
+stages of their accounts. At each stage, the state machine manages the |credential|s
+access to the appropriate browser capabilities.
+
+TODO: update the below image to use ASCII and reflect the algorithms.
```
oh yeah, very much. i think we should move all of our diagrams to ASCII, because these .svgs are really hard to maintain and edit.

i build a ASCII -> SVG renderer that i've been using that i think will be handy to bring to this spec:

https://code.sgo.to/2021/12/07/ascii-diagrams.html

another example:

https://code.sgo.to/2022/04/28/siop.html

npm1 replied:
Is that really easier to maintain? It looks harder to create/update compared with using a diagram editor. I removed the image for now as it is very outdated, TODO still standing.

samuelgoto commented on spec/index.bs:
```diff
@@ -1015,6 +1043,20 @@ dictionary FederatedIdentityProvider {
 The {{CredentialRequestOptions/mediation}} flag is used to determine if the user
 is prompted to select an account.
 
+  :  {{CredentialMediationRequirement/silent}}
```
But lines 1046-1047, 1050-1052 and 1054-1055 does specify what is the algorithm when "silent", "optional" and "required" is passed though, right? 

Can we move this section altogether if we don't use these algorithms yet in implementation?

samuelgoto commented on spec/index.bs:
```diff
@@ -1072,109 +1111,24 @@ When this method is invoked, the user agent MUST execute the following algorithm
+  1. Let |mode| be |options|["{{CredentialRequestOptions/mediation}}"]
+
+  1. TODO(dsinclair): What should {{StateMachine/Session State}} be here?
+  1. If |mode| is "required"
```
Same here. Is this something that we have implemented?

npm1 replied:
Done

samuelgoto commented on spec/index.bs:
```diff
@@ -1237,21 +1191,115 @@ dictionary FederatedTokens {
+        |provider|["{{FederatedIdentityProvider/url}}"]
+    1. Let the |accounts list| be the result of fetching the |accounts_endpoint|
+        with the [=Identity Provider=]'s cookies.  This request MUST NOT follow
+        [[RFC7231#header.location|HTTP redirects]] and instead abort with an
```
Can you add a TODO here to explain why this is the case in the Security section?

samuelgoto reviewed (commented):
On one hand, I still feel like there is a lot in this PR that hasn't been implemented / agreed upon yet, which makes this harder to merge.

Can you take a pass at removing anything that we are already "committed to implementing" so that it meets the bar that we have agreed to (meaning, either it has already been implemented or that we have general agreement that we should - e.g. has passed privacy/security reviews)?

And, on the other hand, there is a lot that *has* been implemented but isn't in this PR. 

It is OK if this PR doesn't cover *everything* that has already been implemented / agreed to, but I feel like it needs to meet the bar for "has only things that we are committed to doing".

The thing that stands out to me are the algorithms for "mode: required/silent/mediated", but I'm guessing there is probably more.

Does that make sense?

npm1 commented:
> On one hand, I still feel like there is a lot in this PR that hasn't been implemented / agreed upon yet, which makes this harder to merge.
> 
> Can you take a pass at removing anything that we are already "committed to implementing" so that it meets the bar that we have agreed to (meaning, either it has already been implemented or that we have general agreement that we should - e.g. has passed privacy/security reviews)?
> 
> And, on the other hand, there is a lot that _has_ been implemented but isn't in this PR.
> 
> It is OK if this PR doesn't cover _everything_ that has already been implemented / agreed to, but I feel like it needs to meet the bar for "has only things that we are committed to doing".
> 
> The thing that stands out to me are the algorithms for "mode: required/silent/mediated", but I'm guessing there is probably more.
> 
> Does that make sense?

I removed what I saw was unimplemented from this PR (although there are still all the examples, which for now I kept). I'm not sure what we have already implemented but is not mentioned on the PR, can you give some concrete pointer? Please take another look.

npm1 commented:
One thing pointed out missing here that is in implementation: the manifest two-tier list

npm1 commented:
I've added the manifest checks!

samuelgoto commented on spec/index.bs:
```diff
@@ -1237,21 +1157,134 @@ dictionary FederatedTokens {
+
+To <dfn>fetch the manifest</dfn> given a {{FederatedIdentityProvider}} |provider|:
+    1. In parallel, perform the following two steps:
+        1. Let |manifestInSet| be the result of running [=check the root manifest=], passing |provider|.
```
thanks!

npm1 replied:
NP!

samuelgoto commented on spec/index.bs:
```diff
@@ -1237,21 +1157,134 @@ dictionary FederatedTokens {
+        1. Let |manifest| be the result of running [=fetch the internal manifest=], passing |provider|.
+    1. If |manifestInSet| is true, return |manifest|, otherwise abort with an error.
+
+To <dfn>check the root manifest</dfn> given a {{FederatedIdentityProvider}} |provider|:
```
Can you add a Note: here somewhere to explain "why" we are doing this (maybe add a link to the github issue?)? Seems hard to infer that from the algorithm itself.

npm1 replied:
Done

samuelgoto commented on spec/index.bs:
```diff
@@ -1237,21 +1157,134 @@ dictionary FederatedTokens {
+    1. Set |rootUrl|'s [=url/path=] to ".well-known/fedcm.json".
+    1. Let |request| be a new <a spec=fetch for=/>request</a> whose [=request/url=] is |rootUrl| and
+        [=request/redirect mode=] set to "error". TODO: what other fields need to be set?
+    1. [=Fetch=] |request| with <a spec=fetch>processResponseConsumeBody</a> set to the following
```
@cbiesinger could you sanity check this algorithm too?

cbiesinger replied:
Just to be clear, I did take a look at this but I haven't fully verified that we get all the interactions with the fetch spec correct

samuelgoto commented on spec/index.bs:
```diff
@@ -1237,21 +1157,134 @@ dictionary FederatedTokens {
+        [[RFC7231#header.location|HTTP redirects]] and instead abort with an
+        error if there are any.
+
+To <dfn>fetch the accounts list</dfn>:
```
maybe move this closer to where it is used?

npm1 replied:
Done

samuelgoto commented on spec/index.bs:
```diff
@@ -1237,21 +1157,134 @@ dictionary FederatedTokens {
+        TODO: explain why we must not follow redirects in the security section.
+    1. Return the |accounts list|
+
+To <dfn noexport>fetch the client metadata</dfn>:
```
same here, move this up?

npm1 replied:
Done

samuelgoto commented on spec/index.bs:
```diff
@@ -272,31 +283,28 @@ could be implemented.
   async function logout() {
     const credential = getFederatedCredential();
-    // This never prompts, rejects the promise in case !credential[“consented”]
+    // This never prompts, rejects the promise in case !credential[“approved”]
```
can we use "registered" and "unregistered" (as in "does the browser think that the user has registered an account in this RP with this IDP?") instead of "approved" or "unapproved"?

npm1 replied:
I renamed but I'm not sure I agree registered is more accurate, as the user agent does not know for sure if the user successfully registered the account, but does know for sure that it approved the RP/IDP communication.

samuelgoto replied:
SGTM. Lets follow up on a separate PR. This has a bunch of good stuff that I think is worth checking in in isoation.

samuelgoto reviewed (commented):
Looking really good @npm1 ! Some minor comments, but otherwise LGTM.

If you resolve these comments, I'll merge, and we can address other comments in other PRs.

Good stuff, this is substantially better than previous revisions! Thanks for limiting the scope of things that we have already implemented / agreed upon, this is really good stuff!

npm1 commented on spec/index.bs:
```diff
@@ -1012,8 +1030,7 @@ dictionary FederatedIdentityProvider {
 
-The {{CredentialRequestOptions/mediation}} flag is used to determine if the user
-is prompted to select an account.
+The {{CredentialRequestOptions/mediation}} flag is currently not used.
```
Done

npm1 commented on spec/index.bs:
```diff
@@ -1072,109 +1090,11 @@ When this method is invoked, the user agent MUST execute the following algorithm
-{{client_id_metadata_endpoint_response/terms_of_service_url}}.
+  1. Let |hint| be |provider|["{{FederatedIdentityProvider/hint}}"]
+  1. Let |credential| be a new {{FederatedCredential}}.
+  1. Set |credential|'s <a>identityProvider</a> to |provider|.
```
Done

npm1 commented on spec/index.bs:
```diff
@@ -1237,21 +1157,134 @@ dictionary FederatedTokens {
+        [=fetch the accounts list=] algorithm.
+    1. Let |account| be the result of running the [=select an account=] from the
+        |accounts list| algorithm.
+    1. If |account| is null, return.
```
It returns null if the user does not select an account, see step 4.

cbiesinger replied:
I think our impl will reject the promise in this case, but we can change that

npm1 replied:
This algorithm is `request consent`, it bails out early if the `account` was not chosen. That means that in the caller above, in step 3 it will reject the promise.

samuelgoto replied:
SGTM

npm1 commented on spec/index.bs:
```diff
@@ -1237,21 +1157,134 @@ dictionary FederatedTokens {
+        to [=approved=].
+    1. Set the account's {{StateMachine/Session State}} from [=logged-out=]
+        to [=logged-in=].
+    1. Set |credential|["id"] to |account|["id"}]
```
Done

cbiesinger commented on spec/index.bs:
```diff
@@ -1237,21 +1156,141 @@ dictionary FederatedTokens {
+        with the [=Identity Provider=]'s cookies.  This request MUST NOT follow
+        [[RFC7231#header.location|HTTP redirects]] and instead abort with an
+        error if there are any.
+        TODO: explain why we must not follow redirects in the security section.
```
Can we link to the accounts endpoint section, which also talks about cookies and referrers?https://fedidcg.github.io/FedCM/#idp-api-accounts-endpoint

npm1 replied:
Done

cbiesinger commented on spec/index.bs:
```diff
@@ -1237,21 +1156,141 @@ dictionary FederatedTokens {
+        header but without the [=Identity Provider=]'s cookies.  This request
+        MUST NOT follow [[RFC7231#header.location|HTTP redirects]] and instead
+        abort with
+        an error if there are any.
```
Again I think we should link to the endpoint section

npm1 replied:
Done

cbiesinger commented on spec/index.bs:
```diff
@@ -1237,21 +1156,141 @@ dictionary FederatedTokens {
+    1. If |accounts list|'s [=list/size=] is 1:
+        1. Let |account| be |accounts list|[0] and return |account|
+    1. If this's {{FederatedCredential/provider}}["accountId"] is provided and
+        an account matches this's {{FederatedCredential/provider}}["accountId"] then return |account|
```
linebreak at 80 characters?

npm1 replied:
Im using 100 but yea this is over.

cbiesinger commented on spec/index.bs:
```diff
@@ -1237,21 +1156,141 @@ dictionary FederatedTokens {
+        1. Set |request|["client_id"] to |provider|["{{FederatedIdentityProvider/clientId}}"].
+        1. Set |request|["nonce"] to |request|["{{FederatedAccountLoginRequest/nonce}}"].
+    1. Let |tokens| be the result of making a POST request to the
+        |manifest|["{{Manifest/id_token_endpoint}}"] with:
```
link to https://fedidcg.github.io/FedCM/#idp-api-id-token-endpoint instead of duplicating the redirect/request body information?

npm1 replied:
Done

cbiesinger commented on spec/index.bs:
```diff
@@ -1237,21 +1156,141 @@ dictionary FederatedTokens {
+        1. Let |manifest| be the result of running [=fetch the internal manifest=], passing |provider|.
+    1. If |manifestInSet| is true, return |manifest|, otherwise abort with an error.
+
+NOTE: We use a two-tier manifest list in order to prevent the [=IDP=] to easily determine the [=RP=]
```
linebreak this entire paragraph at 80 characters?

npm1 replied:
Using 100 chars

cbiesinger commented on spec/index.bs:
```diff
@@ -1237,21 +1156,141 @@ dictionary FederatedTokens {
+    1. Let |url| be |provider|'s {{FederatedIdentityProvider/url}}.
+    1. Let |rootUrl| be a new [=/URL=].
+    1. Set |rootUrl|'s [=url/scheme=] to |url|'s [=url/scheme=].
+    1. Set |rootUrl|'s [=url/host=] to |url|'s [=url/host=].
```
That is incorrect, we set the host to the eTLD+1

npm1 replied:
I think this is the registrable domain, does that look good to you?

cbiesinger replied:
Yes thanks! (Interestingly, reading https://url.spec.whatwg.org/#host-registrable-domain, the registrable domain for something like "github.io" is null... I wonder if that's undesirable for us, though probably doesn't matter in practice)

cbiesinger commented on spec/index.bs:
```diff
@@ -1237,21 +1156,141 @@ dictionary FederatedTokens {
+    1. Set |rootUrl|'s [=url/host=] to |url|'s [=url/host=].
+    1. Set |rootUrl|'s [=url/path=] to ".well-known/fedcm.json".
+    1. Let |request| be a new <a spec=fetch for=/>request</a> whose [=request/url=] is |rootUrl| and
+        [=request/redirect mode=] set to "error". TODO: what other fields need to be set?
```
This is the only part where we link to fetch. I guess you're doing this as a step to converting everything to fetch?

npm1 replied:
Yep. I tried to add the integration properly since I added this one myself, it is left for later to integrate the other fetches :)

cbiesinger commented on spec/index.bs:
```diff
@@ -1283,7 +1322,23 @@ partial interface FederatedCredential {
+  1. Let |promise| be a new {{Promise}}.
+  1. If {{StateMachine/Account State}} is [=unregistered=] or
+      {{StateMachine/Session State}} is [=logged-out=]
+    1. <a for=Promise>Reject</a> |promise| and return |promise|
```
we should eventually specify which exception to reject it with

npm1 replied:
Yep. Added TODO

cbiesinger commented on spec/index.bs:
```diff
@@ -1283,7 +1322,23 @@ partial interface FederatedCredential {
+        |manifest|["{{Manifest/revocation_endpoint}}"] of
+        |provider|["{{FederatedIdentityProvider/url}}"]
+    1. Fetch the |revocation_endpoint| with the [=Identity Provider=]'s
+        cookies.  This request MUST NOT follow
```
again link to the endpoint

npm1 replied:
Done

cbiesinger commented on spec/index.bs:
```diff
@@ -1283,7 +1322,23 @@ partial interface FederatedCredential {
+        cookies.  This request MUST NOT follow
+        [[RFC7231#header.location|HTTP redirects]] and instead abort with an
+        error if there are any.
+    1. If the HTTP response code is not 2xx <a for=Promise>reject</a> |promise| and return |promise|.
```
This is inconsistent with https://fedidcg.github.io/FedCM/#idp-api-revocation-endpoint which specifies 204 (not 2xx)

npm1 replied:
Done (though IDK if this is consistent with the implementation?)

cbiesinger replied:
You are right that we don't enforce that today and that we should make this consistent eventually, one way or another.

cbiesinger reviewed (commented):
Some comments; the most important one probably being the eTLD+1 comment.

samuelgoto commented:
This isn't perfect, but it is a massive clean up that makes a meaningful step forward. I'm going to merge this now because it is substantially better, and we can take a pass at the smaller parts incrementally from this baseline. Thanks @npm1 !
